src/core/instance/inject.js

```diff
@@ -49,7 +49,7 @@ export function resolveInject (inject: any, vm: Component): ?Object {
 
     for (let i = 0; i < keys.length; i++) {
       const key = keys[i]
-      const provideKey = inject[key]
+      const provideKey = inject[key].name
```

Comment:
I'm speculating this will introduce a breaking change for cases where `inject` is an object where the key is the property the injection should be assigned to and the value is the key:
```
const s = Symbol()

const Provider = {
  provide () {
    return {
      [s]: 'foo'
    }
  }
}

const Child = {
  inject: {
    fromProvider: s
  },
  created() {
  	console.log('This is from the provider:', this.fromProvider);
  },
  // ...
}
```

I believe you should be checking if `inject[key]` is a String or Symbol before accessing the `name` property.